### ISSUE TITLE
CLS token mode returns PAD tokens instead of CLS tokens when `padding_side` is `left`

### ISSUE THREAD

**Issue by FremyCompany** [UKPLab/sentence-transformers#3208]
The pooling module assumes the padding side is `right` because it assumes the CLS token is the first token of all embeddings.

https://github.com/UKPLab/sentence-transformers/blob/e2a0098b0fbe10bf9a140a9b1d4c2a3451f1571f/sentence_transformers/models/Pooling.py#L158-L160

This assumption is not correct when `padding_side` is left (e.g. for `LlamaTokenizer`), and this results in the embedding of the `[PAD]` token to be returned instead when sequences have different lengths (or have been padded to a specific multiple of 8).

This is handled properly for `lasttoken`, however.
https://github.com/UKPLab/sentence-transformers/blob/e2a0098b0fbe10bf9a140a9b1d4c2a3451f1571f/sentence_transformers/models/Pooling.py#L212-L234

I think the proper fix would be to properly take into account `padding_side` in both `cls` and `lasttoken` modes, to use the fast approach (`[:,0]`/`[:,-1]`) when the padding side matches the request, and use the attention mask approach otherwise.

But, also, it would probably be good to actually set the proper `padding_side` when tokenizing to match the Pooling module, that would make `lasttoken` faster in most cases. The only time where the slow path is needed is when someone asks a Pooling with both `cls` and `lasttoken` which I don't think will be common.

**Comment by tomaarsen** [comment]
Hello!

Hmm, I imagine it might be tricky to determine what the true CLS token is if the sequence is unbatched, so the `padding_side` would need to be a parameter specified by the user. I don't think there's another way around that, right?

- Tom Aarsen

**Comment by FremyCompany** [comment]
@tomaarsen The padding side has no impact on the results produced by the model. so I think the correct fix is to specify the `right` padding_side in the call to the tokenizer. It's an argument you can pass, and when not passed it defaults to the usual for the class. Given that there are no differences at all between `left` or `right` padding, I think the library can just force-set `right` for now.

**Comment by FremyCompany** [comment]
Probably there:
https://github.com/UKPLab/sentence-transformers/blob/e2a0098b0fbe10bf9a140a9b1d4c2a3451f1571f/sentence_transformers/models/Transformer.py#L506-L512

The documentation is there:
https://huggingface.co/docs/transformers/main_classes/tokenizer#transformers.PreTrained[Tokenizer](https://huggingface.co/docs/transformers/main_classes/tokenizer#transformers.PreTrainedTokenizer.__call__.padding_side).__call__.padding_side

**Comment by tomaarsen** [comment]
I've implemented it here now: https://github.com/huggingface/sentence-transformers/pull/3767
There's still a slight discrepancy between the flattened and non-flattened paths when excluding prompts from the pooling, but I think that would be fine to let be for now. CLS while excluding pooling isn't normally used.

- Tom Aarsen
